### PR TITLE
Intercept Global Installs from npm and Yarn

### LINKED ISSUE #154: intercept global installations
The `yarn` and `npm` shims should intercept requests to do global installations and prevent them with an error by default, where the error message suggests using `notion install` instead. This should be overrideable with an environment variable.

### PR DESCRIPTION
**Pull request by charlespierce**
Closes #154 

Changes
-----
* Added checks to `npm` and `yarn` to intercept global installs and show an error message guiding the user to `notion install` for global packages.
* Allowed this behavior to be overridden by the `NOTION_ALLOW_GLOBAL` environment variable.
* Refactored the `Tool` implementations into their own files within the `tool` subdirectory, but re-exported them from the `tool` module so they can be accessed in the same manner as they were previously.

Notes
-----
* Yarn global installs will always start with `yarn global add`, since `global` is a separate command that must be the first argument, then `add` is the command to install a package.
* npm global installs have a few variations. The command, which is the first argument, can be either `install` or `i`, while the global argument can be anywhere in the argument list and can be either `-g` or `--global`.

Tested
-----
* Installed locally and verified that global installs are intercepted, unless the `NOTION_ALLOW_GLOBAL` environment variable is set.
* Added acceptance tests to verify the interception and non-interception when the env variable is set.

### CODE REVIEW

**Comment by charlespierce** [comment]
Creating the tests turned out to be simpler than I thought, though I did have to work out some kinks with Windows vs Unix handle a `Command` pointed at a missing exectuable.

File: crates/notion-core/src/tool/npm.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        let mut args = args_os();
+        let exe = arg0(&mut args)?;
+
+        if intercept_global_installs() {
+            let mut search_args = args_os().skip(1);
+            let command = search_args.next();
+
+            // npm global installs will be of the form `npm i` or `npm install`
+            // with `-g` or `--global` somewhere in the arguments
+            if (command == Some(OsString::from("i")) || command == Some(OsString::from("install")))
```
[the remark this answers is not in the document]
Comment (dherman) [reply crates/notion-core/src/tool/npm.rs]: Ooh, and the plot thickens: `npm` also accepts `isntall` as an [undocumented alias](https://github.com/npm/npm/issues/2933).

You can find the complete list of `install` aliases [in the source](https://github.com/npm/cli/blob/36682d4482cddee0acc55e8d75b3bee6e78fff37/lib/config/cmd-list.js). I see the following aliases:

- `i`
- `isntall`
- `add`

I know this is icky, but it's what we're signing up for by intercepting global installations. We'll need to track the evolution of `npm` and `yarn` and intercept any new aliases as they come into existence.

(We could get _super_ tricky and cross-reference the alias list for each version of `npm`. But since all we're doing is failing fast, I think it's fine to just intercept the full set of all aliases regardless of what version of `npm` is being used.)

File: crates/notion-core/src/tool/yarn.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+impl Tool for Yarn {
+    fn new(session: &mut Session) -> Fallible<Self> {
+        session.add_event_start(ActivityKind::Yarn);
+
+        let mut args = args_os();
+        let exe = arg0(&mut args)?;
+
+        if intercept_global_installs() {
+            // Yarn global installs must be of the form `yarn global add`
+            if args_os().skip(1).take(2).eq(vec!["global", "add"]) {
```
[the remark this answers is not in the document]
Comment (dherman) [reply crates/notion-core/src/tool/yarn.rs]: More plot-thickening! This doesn't take into account additional flags, e.g. `yarn --verbose global add fnord`. This needs a bit more careful thought to fully characterize the set of legal arguments that can constitute a Yarn global installation.

If it'll take too much work, you can just file a follow-up issue and link to it in a comment. (After all, this is still good progress!)

**Comment by charlespierce** [comment]
@dherman Updated to handle the additional edge cases, and added to the tests to cover those cases as well.
